### PR TITLE
Plot convergence summary and update outdated notebook

### PR DESCRIPTION
**Pull request by FFroehlich**

### CODE REVIEW

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

**Comment by FFroehlich** [comment]
ping @yannikschaelte 

PR travis caching is borked, not sure how to fix that and don't think this is due to this PR.

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

File: pypesto/visualize/optimizer_convergence.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import matplotlib.pyplot as plt
```
Comment (yannikschaelte) [inline pypesto/visualize/optimizer_convergence.py:1]: Would be great to add this to a notebook?

File: pypesto/visualize/optimizer_convergence.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import seaborn as sns
+import numpy as np
+
+
+def optimizer_convergence(result,
+                          ax=None,
+                          size=(18.5, 10.5)):
+    """
+    Scatter plot of function values and gradient values at the end of
+    optimization. Optimizer exit-message is encoded in color. Can help
```
Comment (yannikschaelte) [inline pypesto/visualize/optimizer_convergence.py:12]:
```suggestion
    optimization. Optimizer exit-message is encoded by color. Can help
```

File: pypesto/visualize/optimizer_convergence.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    Returns
+    -------
+
+    ax: matplotlib.Axes
+        The plot axes.
+    """
+    if ax is None:
+        ax = plt.subplots()[1]
+        fig = plt.gcf()
+        fig.set_size_inches(*size)
```
Comment (yannikschaelte) [inline pypesto/visualize/optimizer_convergence.py:38]:
```suggestion
        ax = plt.subplots(figsize=size)[1]
```

File: pypesto/visualize/optimizer_convergence.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+    fvals = result.optimize_result.get_for_key('fval')
+    # ideally use problem.get_reduced_vector here, but checking for nans
```
Comment (yannikschaelte) [inline pypesto/visualize/optimizer_convergence.py:41]: why not use get_reduced vector?

File: pypesto/visualize/optimizer_convergence.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    msgs = result.optimize_result.get_for_key('message')
+    conv_data = pd.DataFrame({
+        'fval': fvals,
+        'grad': grads,
+        'exit message': msgs
+    })
+    sns.scatterplot(x='fval', y='grad', hue='exit message',
+                    data=conv_data, ax=ax)
+    ax.set_yscale('log')
+    ax.set_xscale('log')
```
Comment (yannikschaelte) [inline pypesto/visualize/optimizer_convergence.py:57]: may be good to add default=log arguments as a log scale may not always be ideal?

**Review by yannikschaelte — APPROVED** [review]
Sorry for the delay :+1:

File: pypesto/visualize/optimizer_convergence.py (hunk 4 above)
Comment (FFroehlich) [reply pypesto/visualize/optimizer_convergence.py:41]: see comment, would require passing problem as argument here.

Comment (yannikschaelte) [reply pypesto/visualize/optimizer_convergence.py:41]: Should always hold a result (/result info in the future) https://github.com/ICB-DCM/pyPESTO/blob/master/pypesto/result.py#L201, but fine as is.

**Comment by yannikschaelte** [comment]
> PR travis caching is borked, not sure how to fix that and don't think this is due to this PR.

@FFroehlich where is that the case?

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

**Comment by FFroehlich** [comment]
> > PR travis caching is borked, not sure how to fix that and don't think this is due to this PR.
> 
> @FFroehlich where is that the case?

https://travis-ci.com/github/ICB-DCM/pyPESTO/builds/180572759

seems to be resolved by now

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

**Comment by dweindl** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- pypesto/visualize/optimizer_convergence.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dweindl/pyPESTO/pullRequest?prid=6041703&bid=19796114)

**Pull request merged**